File: src/pathwaytheme/pca/signatures.py
"""Per-observation pathway signatures from a fitted PCA.

Ports pca_analysis.ipynb section 4 / _pca_rollout.run_pca_for_sample:
for each observation, pick its top-N PCs by |score|, then

    signature(obs, pathway) = sum_{k in top PCs} score[obs, k] * loading[k, pathway]

Rank pathways by |signature|, keep the top M (sign preserved: + = up, - = down).
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def compute_signatures(scores_df: pd.DataFrame, loadings_df: pd.DataFrame,
                       top_pcs_per_obs: int, top_pathways: int):
    """Return ``(per_obs, signatures_long)``.

    per_obs : {obs_label: (top_pc_names, top_paths_series)}
    signatures_long : DataFrame [cluster, top_pcs, rank, pathway, signature_score, direction]
    """
    per_obs: dict = {}
    records: list[dict] = []
    for obs_label, s_vec in scores_df.iterrows():
        top_pcs = (s_vec.abs().sort_values(ascending=False)
                   .head(top_pcs_per_obs).index.tolist())
        sig = np.zeros(loadings_df.shape[1])
        for k in top_pcs:
            sig += s_vec[k] * loadings_df.loc[k].values
        sig_ser = pd.Series(sig, index=loadings_df.columns)
        top_idx = sig_ser.abs().sort_values(ascending=False).head(top_pathways).index
        top_paths = sig_ser.loc[top_idx]
        per_obs[obs_label] = (top_pcs, top_paths)
        for rank, (p, v) in enumerate(top_paths.items(), 1):
            records.append({
                "cluster": obs_label,
                "top_pcs": ",".join(top_pcs),
                "rank": rank,
                "pathway": p,
                "signature_score": float(v),
                "direction": "up" if v > 0 else "down",
            })
    return per_obs, pd.DataFrame(records)


def top_pathways_per_pc(loadings_df: pd.DataFrame, var_expl: np.ndarray,
                        top_n: int) -> pd.DataFrame:
    """Long table of the top-|loading| pathways per PC (ports the TSV logic)."""
    pc_names = list(loadings_df.index)
    rows: list[dict] = []
    for i, k in enumerate(pc_names):
        lv = loadings_df.loc[k]
        top = lv.abs().sort_values(ascending=False).head(top_n).index
        for rank, p in enumerate(top, 1):
            rows.append({
                "PC": k,
                "variance_explained": float(var_expl[i]),
                "rank": rank,
                "pathway": p,
                "loading": float(lv[p]),
                "direction": "up" if lv[p] > 0 else "down",
            })
    return pd.DataFrame(rows)

File: r/R/signatures.R
## Per-observation pathway signatures from a fitted PCA.
##
## For each observation, pick its top-N PCs by |score|, then
##
##     signature(obs, pathway) = sum_{k in top PCs} score[obs, k] * loading[k, pathway]
##
## Rank pathways by |signature|, keep the top M (sign preserved: + = up,
## - = down).

#' Per-observation pathway signatures
#'
#' @param scores observations x PCs matrix.
#' @param loadings PCs x pathways matrix.
#' @param top_pcs_per_obs PCs combined into one observation's signature.
#' @param top_pathways pathways kept per observation.
#' @return A list with `per_obs` (a list of `top_pcs` / `values` per
#'   observation) and `long` (a data.frame with `cluster`, `top_pcs`, `rank`,
#'   `pathway`, `signature_score`, `direction`).
#' @export
pt_compute_signatures <- function(scores, loadings, top_pcs_per_obs,
                                  top_pathways) {
  pc_names <- colnames(scores)
  pathways <- colnames(loadings)
  obs_labels <- rownames(scores)

  per_obs <- vector("list", nrow(scores))
  names(per_obs) <- obs_labels
  records <- vector("list", nrow(scores))

  for (i in seq_len(nrow(scores))) {
    s_vec <- scores[i, ]
    top_pcs <- pc_names[utils::head(.pt_order_desc(abs(s_vec)), top_pcs_per_obs)]
    sig <- as.vector(s_vec[top_pcs] %*% loadings[top_pcs, , drop = FALSE])
    names(sig) <- pathways

    keep <- utils::head(.pt_order_desc(abs(sig)), top_pathways)
    top_paths <- sig[keep]
    per_obs[[i]] <- list(top_pcs = top_pcs, values = top_paths)
    records[[i]] <- data.frame(
      cluster = obs_labels[[i]],
      top_pcs = paste(top_pcs, collapse = ","),
      rank = seq_along(top_paths),
      pathway = names(top_paths),
      signature_score = as.double(top_paths),
      direction = ifelse(top_paths > 0, "up", "down"),
      stringsAsFactors = FALSE)
  }

  long <- if (length(records)) do.call(rbind, records) else
    data.frame(cluster = character(), top_pcs = character(), rank = integer(),
               pathway = character(), signature_score = double(),
               direction = character(), stringsAsFactors = FALSE)
  rownames(long) <- NULL
  list(per_obs = per_obs, long = long)
}

#' Top-loading pathways per principal component
#'
#' @param loadings PCs x pathways matrix.
#' @param variance_explained variance explained per PC.
#' @param top_n pathways tabulated per component.
#' @return A long data.frame with `PC`, `variance_explained`, `rank`,
#'   `pathway`, `loading`, `direction`.
#' @export
pt_top_pathways_per_pc <- function(loadings, variance_explained, top_n) {
  pc_names <- rownames(loadings)
  rows <- vector("list", length(pc_names))
  for (i in seq_along(pc_names)) {
    lv <- loadings[i, ]
    keep <- utils::head(.pt_order_desc(abs(lv)), top_n)
    rows[[i]] <- data.frame(
      PC = pc_names[[i]],
      variance_explained = as.double(variance_explained[[i]]),
      rank = seq_along(keep),
      pathway = colnames(loadings)[keep],
      loading = as.double(lv[keep]),
      direction = ifelse(lv[keep] > 0, "up", "down"),
      stringsAsFactors = FALSE)
  }
  out <- do.call(rbind, rows)
  rownames(out) <- NULL
  out
}
